src/nccl_ofi_idpool.cpp

```diff
@@ -4,116 +4,53 @@
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_init(nccl_ofi_idpool_t *idpool, size_t size)
-{
-	int ret = 0;
-
-	assert(NULL != idpool);
-
-	if (0 == size) {
-		/* Empty or unused pool */
-		idpool->ids = NULL;
-		idpool->size = 0;
-		return ret;
-	}
 
-	/* Scale pool size to number of 64-bit uints (rounded up) */
-	size_t num_long_elements = NCCL_OFI_ROUND_UP(size, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
-
-	/* Allocate memory for the pool */
-	idpool->ids = (uint64_t *)malloc(sizeof(uint64_t) * num_long_elements);
-
-	/* Return in case of allocation error */
-	if (NULL == idpool->ids) {
-		NCCL_OFI_WARN("Unable to allocate ID pool");
-		return -ENOMEM;
+nccl_ofi_idpool_t::nccl_ofi_idpool_t(const size_t& size) : 
+	size_(size)
+{
+	if (size_ == 0) {
+		idpool_ = std::vector<uint64_t>();
+		return;
 	}
 
-	/* Set all IDs to be available */
-	memset(idpool->ids, 0xff, size / 8);
-	if (size % 8) {
-		idpool->ids[num_long_elements - 1] = (1ULL << (size % (sizeof(uint64_t) * 8))) - 1;
-	}
+	size_t num_long_elements = NCCL_OFI_ROUND_UP(size_, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
+	idpool_ = std::vector<uint64_t>(num_long_elements, 0xffffffffffffffff);
 
-	/* Initialize mutex */
-	ret = nccl_net_ofi_mutex_init(&idpool->lock, NULL);
-	if (OFI_UNLIKELY(ret)) {
-		NCCL_OFI_WARN("Unable to initialize mutex");
-		free(idpool->ids);
-		idpool->ids = NULL;
-		return ret;
+	if (size_ % 8 != 0) {
+		idpool_.at(num_long_elements - 1) = (1ULL << (size_ % (sizeof(uint64_t) * 8))) - 1;
 	}
-
-	idpool->size = size;
-
-	return ret;
 }
 
-/*
- * @brief	Allocate an ID
- *
- * Extract an available ID from the ID pool, mark the ID as
- * unavailable in the pool, and return extracted ID. No-op in case
- * no ID was available.
- *
- * This operation is locked by the ID pool's internal lock.
- *
- * @param	idpool
- *		The ID pool
- * @return	the extracted ID (zero-based) on success,
- *		negative value on error
- */
-int nccl_ofi_idpool_allocate_id(nccl_ofi_idpool_t *idpool)
+
+int nccl_ofi_idpool_t::allocate_id()
 {
-	assert(NULL != idpool);
+	std::lock_guard<std::mutex> l(lock);
 
-	if (0 == idpool->size) {
+	if (0 == size_) {
 		NCCL_OFI_WARN("Cannot allocate an ID from a 0-sized pool");
 		return -ENOMEM;
 	}
-
-	if (OFI_UNLIKELY(NULL == idpool->ids)) {
-		NCCL_OFI_WARN("Invalid call to nccl_ofi_allocate_id with uninitialized pool");
-		return -EINVAL;
-	}
-
-	/* Scale pool size to number of 64-bit uints (rounded up) */
-	size_t num_long_elements = NCCL_OFI_ROUND_UP(idpool->size, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
-
-	nccl_net_ofi_mutex_lock(&idpool->lock);
-
 	int entry_index = 0;
 
 	bool found = false;
 	size_t id = 0;
-	for (size_t i = 0; i < num_long_elements; i++) {
-		entry_index = __builtin_ffsll(idpool->ids[i]);
+
+	/* Iterate over each of the idpool_ vector elements */
+	for (size_t i = 0; i < idpool_.size(); i++) {
+		entry_index = __builtin_ffsll(idpool_.at(i));
```

Comment:
it's a vector.  I'd just write this and line 53 below as `idpool[i]`.

Reply:
Fixed in force push. I had originally used `.at` since it has built-in bound-checking, but it would make sense with our current exception handling that we would rather then the bounds ourselves instead of throwing an exception.